### PR TITLE
AutoScaling: support Managed User Data

### PR DESCRIPTION
**Pull request by weizhouapache**
### Description

This PR fixes the issue that template userdata and userdata policy are not considered in the AutoScale VM group.

Related doc PR: https://github.com/apache/cloudstack-documentation/pull/335

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [x] Major
- [ ] Minor
- [ ] Trivial

### CODE REVIEW

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan package

File: api/src/main/java/org/apache/cloudstack/api/BaseCmd.java [hunk 1]
```diff
@@ -456,4 +459,18 @@ public ApiCommandResourceType getApiResourceType() {
...
+            Iterator iter = parameterCollection.iterator();
+            while (iter.hasNext()) {
+                HashMap<String, String> value = (HashMap<String, String>)iter.next();
+                for (Map.Entry<String,String> entry: value.entrySet()) {
+                    detailsMap.put(entry.getKey(),entry.getValue());
+                }
+            }
+        }
+        return detailsMap;
+    }
```
Comment (DaanHoogland) [inline api/src/main/java/org/apache/cloudstack/api/BaseCmd.java:475]: Good refactor! This is happening so often. Did you replace all occurences?

**Review by DaanHoogland — APPROVED** [review]
clgtm, ...

**Comment by weizhouapache** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan test matrix

File: api/src/main/java/org/apache/cloudstack/api/BaseCmd.java (hunk 1 above)
Comment (weizhouapache) [reply api/src/main/java/org/apache/cloudstack/api/BaseCmd.java:475]: updated 2 more APIs

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test matrix

**Comment by weizhouapache** [comment]
@blueorangutan test

**Comment by weizhouapache** [comment]
@blueorangutan test rocky8 kvm-rocky8

**Review by harikrishna-patnala — APPROVED** [review]
LGTM

Verified few scenarios by creating VM auto profiles having userdata overridden policies with registered userdata and also with manual userdata

**Comment by borisstoyanov** [comment]
@blueorangutan test centos7 vmware-67u3

**Comment by borisstoyanov** [comment]
@weizhouapache could you please see the merge conflicts.

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan test matrix

**Review by rajujith — APPROVED** [review]
LGTM. I Verified that managed user data works fine with Autoscale VM groups.

**Comment by weizhouapache** [comment]
thanks @harikrishna-patnala @DaanHoogland for code review
thanks @borisstoyanov @rajujith for testing

merging

**Pull request merged**